### ISSUE TITLE
Doc Review: Security Options

### ISSUE THREAD

**Issue by vburenin** [vmware/vic#3329]
--tls-cname - might not be clear for some users what this actually is. I would add an additional text that says - this is a FQDN of a host that is going to use this certificate. 

Update: There is a text saying what this is way bellow. We should move it to the top and rephrase it to a little simple language.

**Comment by vburenin** [comment]
Common Name cannot be in the form of "vic*.example.net", Golang doesn't implement such type of pattern, we should mention it. 

- vic.*.example.net is allowed.
- vic*.example.net is not allowed.

**Comment by hickeng** [comment]
Correction for @vburenin 's comment: Common Name **cannot** be in the form `vic.*.example.net` - from the RFC:
```
- A "*" wildcard character MAY be used as the left-most name
     component in the certificate.  For example, *.example.com would
     match a.example.com, foo.example.com, etc. but would not match
     example.com.
```

@stuclem I've added the section below as I think the phrasing is more inline with the way people will think about their needs. I did not duplicate the details about cert path or the specifics of where generated certificates get placed as those are already correct.

## Security Options ##

As a convenience `vic-machine create` provides the option of generating a client certificate, server certificate, and certificate authority (CA) as appropriate when deploying a VCH. The generated certificates are functional, but do not allow for fine control over aspects such as expiration, intermediate certificate authorities, et al. To exercise fine control over the certificates used, obtain or generate the certificates ahead of time and use the `--key`, `--cert`, and `--tls-ca` options to pass them to `vic-machine create`.

### Restrict access to the Docker API ###
Authentication of API clients is done using client certificates with this configuration frequently being referred to as `tlsverify` in documentation dealing with containers and Docker. A client certificate is accepted if it is signed by a CA provided via one or more instances of the `--tls-ca` option. In the case of the generated certificates, a certificate authority is created and used to create and sign a single client certificate.

When using the docker client it also validates the server either via CAs that are present in the client system root certificate bundle, or explicitly provided via the `--tlscacert` docker client option. As part of this validation, the _server_ certificate must explicitly state at least one of the following, and must match the name or address the client uses to access the server:
a. the FQDN used to communicate with the server
b. the IP address used to communicate with the server
c. a wildcard domain which will match all FQDNs in a specific subdomain ([example](https://en.wikipedia.org/wiki/Wildcard_certificate#Example))

#### `--tls-cname` ####
This option is used to supply the FQDN or IP address that should be embedded in the generated _server_ certificate. This is one of the three types listed above; FQDN, IP, or domain wildcard. If providing a server certificate directly (via the `--cert` option), then this is used as a sanity check to ensure the certificate is valid for the deployment.

If this option is omitted and a static IP address is configured for the _client_ interface, that address is used, with the same result as if `--tls-cname=x.x.x.x` had been specified.

#### `--tls-ca` ####
This option (short for `--ca`) may be specified multiple times and should point to a file containing the _public_ portion of a certificate authority. These authorities are then used to validate client certificates offered as credentials for API access.

### Unrestricted access to the Docker API ###
To deploy a VCH that does not restrict access to the Docker API, you should use the `--no-tlsverify` option. This option prevents the use of certificate authorities for _client authentication_, however a server certificate is still required. A custom server certificate may be supplied with the `--cert` and `--key` option, otherwise a self-signed server certificate will be generated.

**Comment by vburenin** [comment]
@hickeng There are different implementation how wildcard is treated. The GoLang one is the strict one that follows RFC. I saw confusions when people switching from python to golang had problems with hostname verifications. https://tools.ietf.org/html/rfc6125#section-7.2 actually talks about it.

**Comment by stuclem** [comment]
Thank you @vburenin and @hickeng !

I have taken @hickeng's text, but reworked it a bit for stylistic reasons. I also retained some of my original content where I thought that it was valid to do so. Thank you George for such a detailed rewrite.

If the debate about wildcards, please continue the debate here, if necessary.

Changes are in https://github.com/stuclem/vic/commit/eae0423b2c7923aec410a292b3e8f13f0cd02f6f.

Updated section is [here](https://github.com/stuclem/vic/blob/master/doc/user_doc/vic_installation/vch_installer_options.md#security).

This most definitely needs a second pass.

**Comment by dougm** [comment]
LGTM @stuclem .  May want to mention the directory/file permissions for the generated cert/key are readable only by the user who created them.

**Comment by stuclem** [comment]
Thanks @dougm. Done. Change is in https://github.com/stuclem/vic/commit/518bedd1537a8ddb668fd052185c9773d7d528c1.

If I could get an LGTM from @vburenin and @hickeng, I'd be delighted.

Thanks!

**Comment by vburenin** [comment]
LGTM @stuclem

**Comment by hickeng** [comment]
@vburenin while there's debate about what wildcards _should_ mean, we need to document how they actually function in the Go ecosystem and that's the strict RFC.

@stuclem minor comments and corrections.

**Comment by stuclem** [comment]
Thanks @vburenin and @hickeng. @hickeng I implemented your comments from https://github.com/stuclem/vic/commit/eae0423b2c7923aec410a292b3e8f13f0cd02f6f. 

Also added a new section "Connecting to the VCH" in the [Obtain a VCH](https://github.com/stuclem/vic/blob/master/doc/user_doc/vic_app_dev/obtain_vch.md) topic in the App Dev doc, which takes the info that you provided about which Docker options are required with each type of TLS auth when you run Docker commands.

Updates are in https://github.com/stuclem/vic/commit/5440ffd335c9a0ae9b7cd08e89b0f8abb5aa051a.

If you could just take a look at the Connecting to the VCH section, then I think that we'll be done. 

Thanks again!

Think that we can close this one now. Thanks both!

**Comment by stuclem** [comment]
I took the Connecting to the VCH info from @hickeng 's comments pretty much verbatim, so it's safe to close this.
